azurecosmos/src/main/java/site/ycsb/db/AzureCosmosClient.java

```diff
@@ -187,218 +245,251 @@ private int getIntProperty(String propertyName, int defaultValue) {
-        e = null;
+        CosmosItemResponse<ObjectNode> response = container.readItem(key, new PartitionKey(key), ObjectNode.class);
+        readEtag = response.getETag();
+        ObjectNode node = response.getItem();
+
+        for (Entry<String, ByteIterator> pair : values.entrySet()) {
+          node.put(pair.getKey(), pair.getValue().toString());
+        }
+
+        CosmosItemRequestOptions requestOptions = new CosmosItemRequestOptions();
+        requestOptions.setIfMatchETag(readEtag);
+        PartitionKey pk = new PartitionKey(key);
+        container.replaceItem(node, key, pk, requestOptions);
+
+        return Status.OK;
+      } catch (CosmosException e) {
+        if (!AzureCosmosClient.includeExceptionStackInLog) {
+          e = null;
+        }
+        LOGGER.error("Failed to update key {} to collection {} in database {} on attempt {}", key, table,
+            AzureCosmosClient.databaseName, attempt, e);
+
+        // Azure Cosmos DB does not have patch support. Until then we need to read
+        // the document, update in place, and then write back.
+        // This could actually be made more efficient by using a stored procedure
+        // and doing the read/modify write on the server side. Perhaps
+        // that will be a future improvement.
       }
-      LOGGER.error("Failed to update key {}", key, e);
-      return Status.ERROR;
     }
-  
-    return Status.OK;
+
+    return Status.ERROR;
   }
 
+  /**
+   * Insert a record in the database. Any field/value pairs in the specified
+   * values HashMap will be written into the record with the specified record key.
+   *
+   * @param table  The name of the table
+   * @param key    The record key of the record to insert.
+   * @param values A HashMap of field/value pairs to insert in the record
+   * @return Zero on success, a non-zero error code on error
+   */
   @Override
   public Status insert(String table, String key, Map<String, ByteIterator> values) {
-    Document documentDefinition = getDocumentDefinition(key, values);
-    ResourceResponse<Document> resourceResponse = null;
-    RequestOptions requestOptions = getRequestOptions(key);
-  
+    if (LOGGER.isDebugEnabled()) {
+      LOGGER.debug("Insert key: " + key + " into table: " + table);
+    }
+
     try {
-      if (this.useUpsert) {
-        resourceResponse = AzureCosmosClient.client.upsertDocument(getDocumentCollectionLink(this.databaseName, table),
-            documentDefinition,
-            requestOptions,
-            true);
+      CosmosContainer container = AzureCosmosClient.database.getContainer(table);
+      PartitionKey pk = new PartitionKey(key);
+      ObjectNode node = OBJECT_MAPPER.createObjectNode();
+
+      node.put("id", key);
+      for (Map.Entry<String, ByteIterator> pair : values.entrySet()) {
+        node.put(pair.getKey(), pair.getValue().toString());
+      }
+      if (AzureCosmosClient.useUpsert) {
+        container.upsertItem(node);
```

Comment:
please pass id and pk here that helps to prevent parsing the content by the sdk to discover the id/pk.

```suggestion
container.upsertItem(id, pk, node);
```

https://azuresdkdocs.blob.core.windows.net/$web/java/azure-cosmos/latest/com/azure/cosmos/CosmosContainer.html#upsertItem-T-com.azure.cosmos.models.PartitionKey-com.azure.cosmos.models.CosmosItemRequestOptions-

Reply:
this will have perf impact.